### DISCUSSION TITLE
Homebrew supported use of pyenv and Python is what?

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#2138] [Everyday usage]
A lot of stuff out there (mostly older possibly outdated information) on how to use `pyenv` to manage Python on macOS.
`pyenv` states it's a breeze to install via Homebrew, but how does Homebrew expect users to install Python, using `brew install python` or using `pyenv install python 3.9.7`?

If we choose `pyenv` as the installer for Python will Homebrew work dependably?

Looking at the upcoming changes to Python 3.12,  it seems safer to use `pyenv` to install Python versions.

I have spent about 5 hours on this so giving up and hoping for some help.

**Comment by alebcay** [comment]
> how does Homebrew expect users to install Python, using `brew install python` or using `pyenv install python 3.9.7`?

Either should work.

> If we choose `pyenv` as the installer for Python will Homebrew work dependably?

Yes. Note that packages in Homebrew that depend on Python will always use Homebrew's version of Python. So you may end up with both copies on your machine anyways.

Generally speaking - if you're depending on the version of Python you install to not change until you want it to (e.g. for development purposes), I would advise against using Homebrew Python. It exists more to serve Homebrew/Homebrew packages that need it, and provide a REPL/sandbox once in a while.

See also - [Homebrew Python is not for you](https://justinmayer.com/posts/homebrew-python-is-not-for-you/).

**Reply by ghost** [reply]
Thank you. I am not using for development. I have just read that the pyenv install method is superior to installing python via homebrew. I just want to confirm because I am getting one error when testing python via homebrew.

```
$ brew test python
==> Testing python@3.9
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import sqlite3
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -m venv /private/tmp/pythonA3.9-test-20210915-51653-1bxn8vr/myvenv
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import _ctypes
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import _decimal
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import _gdbm
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import pyexpat
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -c import zlib
==> /usr/local/Cellar/python@3.9/3.9.7/bin/python3.9 -Sc 'import tkinter' 2>&1
Error: python@3.9: failed
An exception occurred within a child process:
  Minitest::Assertion: Expected: 1
  Actual: 0
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.14.4/lib/minitest/assertions.rb:183:in `assert'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.14.4/lib/minitest/assertions.rb:218:in `assert_equal'
/usr/local/Homebrew/Library/Homebrew/formula_assertions.rb:65:in `shell_output'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python@3.9.rb:469:in `block in <class:PythonAT39>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2091:in `block (3 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2090:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:943:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2089:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2342:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2083:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:43:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:48:in `<main>'
```

**Reply by alebcay** [reply]
Homebrew Python has two packages `python@3.9` and `python-tk@3.9`. This split was made so that people installing just Python don't also need to have a `tcl-tk` dependency unless they want it. This check that failed verifies that the `tkinter` module is not present (it shouldn't be, because we've stripped it out of the `python@3.9` package and put it in `python-tk@3.9`.

I would advise trying to reinstall `python@3.9` and see if that fixes the issue, or try installing and then uninstalling `python-tk@3.9`. If the issue persists, my guess is that maybe the Homebrew Python is picking up the `tkinter` module from a different Python installation, in which case you should check your `PYTHONPATH`, etc. (make sure you try running the `brew test` when your `pyenv` is not activated).

We're working on improving how Homebrew Python works in https://github.com/Homebrew/homebrew-core/issues/76621. Part of this is also waiting for Python/pip devs to decide on how to do some things for all distributions.

**Reply by ghost** [reply]
Thank you for the super detailed explanation - I couldn't find anything related to this.

I'm sorry for the confusion, I don't have python installed via homebrew any longer. I removed it and installed python with `pyenv` yesterday.  I was unaware of this information.

The `tikinter` library is a part of the python install and is installed at: `/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter`

```
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/__init__.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/__main__.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/__pycache__
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/colorchooser.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/commondialog.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/constants.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/dialog.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/dnd.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/filedialog.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/font.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/messagebox.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/scrolledtext.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/simpledialog.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/test
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/tix.py
/Users/john/.pyenv/versions/3.9.7/lib/python3.9/tkinter/ttk.py
```

I removed homebrew python because I didn't want duplicated python installs. I also have been seeing some python warnings about how brew has installed python. I installed it with `pyenv` and symlinked into  `/usr/local/Cellar/python@3.9/3.9.7`

I don't have `python-tk@3.9` installed in `/usr/local/Cellar`

I don't know how many tests have failed here or if testing was supposed to continue after this failure. It's not clear that other tests are being run after this that are passing. It looked like a total failed test not just one issue.

I tried to manually remove the `tkinter` library from my non-brew python install but that doesn't work. Didn't prevent this failed test

Is there a way to install `python` using `pyenv` without the `tkinter` library? I guess that would have to be using the package using flags - yuck.

**Reply by alebcay** [reply]
Yes, at this point it would probably be easier for you to just use `pyenv`'s Python and whatever software you use in Homebrew, you install outside of Homebrew instead and somehow make it point to `pyenv`'s Python.

Removing the `tkinter` library from `pyenv`'s Python is probably non-trivial or gross as you mentioned.

> I removed homebrew python because I didn't want duplicated python installs. I also have been seeing some python warnings about how brew has installed python. 

FWIW this is related to deprecated behavior coming up in Python 3.10/recent versions of `pip`. The `pip` team decided to just add that warning message, Homebrew team has no way of suppressing it.

Homebrew team is aware that of the deprecation upstream, but so far there has yet to be a finalized resolution method provided from Python/pip upstream. So Homebrew team is in a bit of a weird spot where users are getting this warning but maintainers don't really have any actionable fix.

In any case, the breakage will occur in 3.12 so we're expecting to resolve this before then. This entire deprecation message is discussed/tracked in https://github.com/Homebrew/homebrew-core/issues/76621.

**Reply by ghost** [reply]
A few brew packages are failing `brew test` and some processes are crashing when run. It looks like I need to create a symlink for certain frameworks to fix this... will research.

```
Error: mkvtomp4: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/mkvtomp4/2.0_2/bin/mkvtomp4 --help
```

```
Error: sphinx-doc: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/sphinx-doc/4.2.0/bin/sphinx-quickstart -pPorject -aAuthor -v1.0 -q /private/tmp/sphinx-doc-test-20210915-81047-12xiwy4
/usr/local/Homebrew/Library/Homebrew/formula.rb:2291:in `block in system'
```

```
Error: streamlink: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/streamlink/2.4.0/bin/streamlink https://vimeo.com/189776460 360p -o video.mp4
/usr/local/Homebrew/Library/Homebrew/formula.rb:2291:in `block in system'
```

```
Error: vapoursynth: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/vapoursynth/54/bin/vspipe --version
/usr/local/Homebrew/Library/Homebrew/formula.rb:2291:in `block in system'
```

```
Error: youtube-dl: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/youtube-dl/2021.6.6/bin/youtube-dl --simulate https://www.youtube.com/watch?v=pOtd1cbOP7k
/usr/local/Homebrew/Library/Homebrew/formula.rb:2291:in `block in system'
```

```
Error: mpv: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/mpv/0.33.1/bin/mpv --ao=null /usr/local/Homebrew/Library/Homebrew/test/support/fixtures/test.wav
/usr/local/Homebrew/Library/Homebrew/formula.rb:2291:in `block in system'
```

crash log for mpv
```
Dyld Error Message:
  dyld: Using shared cache: F0A21792-ED7E-33A3-A0DD-24B063378C1E
Library not loaded: /usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/Python
  Referenced from: /usr/local/opt/vapoursynth/lib/libvapoursynth-script.0.dylib
  Reason: image not found
```

**Reply by ghost** [reply]
Where would I find the brew configuration information for building python for the brew install?

**Reply by carlocab** [reply]
Do you mean how Homebrew configures Python3? You can find it here: https://github.com/Homebrew/homebrew-core/blob/e79d8126370187be3bd7be6bdbffacf803732113/Formula/python@3.9.rb

It seems to me a lot easier to live with a second Python installation (you actually have two more provided by your system...) than try to make the Python you've installed with `pyenv` replace Homebrew Python3.
